Enable update in config between checkpoint and restore for session cache

anjumfatima90 opened the pull request:
for #26845

anjumfatima90 committed:
Enable update in config between checkpoint and restore for sessionCache

Wait for config changes to complete so changes are visible to service components.

anjumfatima90 committed:
Test update in config between checkpoint and restore for sessionCache

anjumfatima90 committed:
Use the addMultiThreadedHook method to register the hook. Use the MIN rank so that it runs earlier.

anjumfatima90 commented:
#build

anjumfatima90 commented:
#build

tjwatson commented on dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/ConfigRefresher.java:
```diff
@@ -82,12 +79,7 @@ public class ConfigRefresher {
         metatypeTracker.open();
 
         changesEndedHook = new ChangesEndedHook();
```
We don't need to keep a reference to this hook after it is added.  Could just do:

```java
        CheckpointPhase.getPhase().addMultiThreadedHook(Integer.MIN_VALUE, new ChangesEndedHook());
```

tjwatson commented on dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java:
```diff
@@ -149,7 +149,7 @@ private void cacheInit() {
         try {
             if (cacheStoreService.cacheManager == null)
-                cacheStoreService.activateLazily();
+                cacheStoreService.activateLazily();              
```
It would be nice to avoid adding unnecessary whitespace.  Can make this change without doing another build IMO.

tjwatson commented on dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMap.java:
```diff
@@ -173,7 +173,7 @@ private void cacheInit() {
             if (trace && tc.isDebugEnabled())
                 tcInvoke(cacheStoreService.tcCacheManager, "getCache", metaCacheName, "String", "ArrayList");
-
+            
```
Same whitespace comment.

tjwatson commented on dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheStoreService.java:
```diff
@@ -541,7 +548,7 @@ public void setCompletedPassivation(boolean isInProcessOfStopping) {
     }
 
-    protected void setLibrary(Library library) {
+    protected void setLibrary(Library library) {        
```
whitespace.

anjumfatima90 committed:
Add Activator to manage config condition.

For regular server run the change in config would recycle the app. This is ensured by the Activator which listens to updated config event and re-registers the config condition service.
For checkpoint/restore case, any change in config would go through the modified method of the component.

Co-authored-by: Thomas Watson <tjwatson@us.ibm.com>

anjumfatima90 commented:
#build

anjumfatima90 commented:
#build

anjumfatima90 commented:
#build

dazavala reviewed (dismissed):
Changes look good, and for the most part I understand them.  I'm not quite clear whether `CheckpointImpl.waitForConfig()` is now vestigial in light of adding  `notification.waitForCompletion()` to `ConfigRefresher.changesEndedHook.restore()`.  Will follow up with you.

anjumfatima90 committed:
Remove waitForConfig method and the RuntimeUpdateListener since we wait for config updates in the changesEnded
Hook inside ConfigRefresher

anjumfatima90 commented:
#build

dazavala commented on dev/io.openliberty.checkpoint/src/io/openliberty/checkpoint/internal/CheckpointImpl.java:
```diff
@@ -297,7 +294,6 @@ public void notificationCreated(RuntimeUpdateManager updateManager, RuntimeUpdat
         if (RuntimeUpdateNotification.CONFIG_UPDATES_DELIVERED.equals(notification.getName()) && jvmRestore.compareAndSet(true, false)) {
             debug(tc, () -> "Processing config on restore.");
             final CountDownLatch localLatch = new CountDownLatch(1);
```
- There's no need for `localLatch`.  The latch is seeded with value `1` and is then used by `waitForConfig()` to block the checkpoint thread until the either the latch is decremented to `0`, times out, or the thread is interrupted,  You removed `waitForConfig()` -- i.e. you removed the wait on the latch -- so there's no need to call `localLatch.countDown()` in the `CompletionListener` methods. 
- I like that the `CompletionListener` methods provide trace that define the window of the restore config update.  If that's not a concern, then you could remove the `RuntimeUpdateListener` support from `CheckpointImpl`.

anjumfatima90 replied:
Removed, thanks!

dazavala replied:
Looks good.  Thank you @anjumfatima90.

dazavala reviewed (changes requested):
@anjumfatima90 I requested further changes to `CheckpointImpl` for commit 72f7eaec4d65c59036aee58ab09d4305fcdb15ed.  Thank you.

anjumfatima90 commented:
#build

dazavala reviewed (dismissed):
Changes in recent commit 4ef6927387d70559eb738f2af26a16c303079b86 LGTM.

SmithaSubbarao reviewed (commented):
My high level understanding of these changes is that a new class, Activator is being used to track updates (to the config) for storing/caching  data from users' applications and components during a checkpoint and restore (as opposed to updating notifications after a restore). I'm confused on how the checkpoint hook changes in this PR so I will need to follow up about this with @anjumfatima90 and @dazavala .

anjumfatima90 commented:
#build

tjwatson commented:
#libby

anjumfatima90 commented:
#build

uitesterhero commented:
This pr has caused intermittent but often failures in two buckets and will need to be reverted - see defect 299821[https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.workitem.viewWorkItem&id=299821](url)
